src/main/fc/rc.c

```diff
@@ -524,38 +523,142 @@ static FAST_CODE void processRcSmoothingFilter(void)
     for (int i = 0; i < PRIMARY_CHANNEL_COUNT; i++) {
         float *dst = i == THROTTLE ? &rcCommand[i] : &setpointRate[i];
         if (rcSmoothingData.filterInitialized) {
-            *dst = pt3FilterApply(&rcSmoothingData.filter[i], rxDataToSmooth[i]);
+            *dst = pt3FilterApply(&rcSmoothingData.filterSetpoint[i], rxDataToSmooth[i]);
         } else {
             // If filter isn't initialized yet, as in smoothing off, use the actual unsmoothed rx channel data
             *dst = rxDataToSmooth[i];
         }
-    }
 
-    // for ANGLE and HORIZON, smooth rcDeflection on pitch and roll to avoid setpoint steps
-    bool smoothingNeeded = (FLIGHT_MODE(ANGLE_MODE) || FLIGHT_MODE(HORIZON_MODE)) && rcSmoothingData.filterInitialized;
+    }
     for (int axis = FD_ROLL; axis <= FD_YAW; axis++) {
-        if (smoothingNeeded && axis < FD_YAW) {
-            rcDeflectionSmoothed[axis] = pt3FilterApply(&rcSmoothingData.filterDeflection[axis], rcDeflection[axis]);
+        // Feedforward smoothing
+        feedforwardSmoothed[axis] = pt3FilterApply(&rcSmoothingData.filterFeedforward[axis], feedforwardRaw[axis]);
+        // Horizon mode smoothing of rcDeflection on pitch and roll to provide a smooth angle element
+        const bool smoothRcDeflection = FLIGHT_MODE(HORIZON_MODE) && rcSmoothingData.filterInitialized;
+        if (smoothRcDeflection && axis < FD_YAW) {
+            rcDeflectionSmoothed[axis] = pt3FilterApply(&rcSmoothingData.filterRcDeflection[axis], rcDeflection[axis]);
         } else {
             rcDeflectionSmoothed[axis] = rcDeflection[axis];
         }
     }
-
 }
 #endif // USE_RC_SMOOTHING_FILTER
 
+FAST_CODE_NOINLINE void initAveraging (void)
+{
+    feedforwardAveraging = pidGetFeedforwardAveraging();
+    for (int i = 0; i < XYZ_AXIS_COUNT; i++) {
+        laggedMovingAverageInit(&feedforwardDeltaAvg[i].filter, feedforwardAveraging + 1, (float *)&feedforwardDeltaAvg[i].buf[0]);
+    }
+}
+
+FAST_CODE_NOINLINE void calculateFeedforward (int axis)
+{
+    const float rxInterval = currentRxRefreshRate * 1e-6f; // seconds
+    const float rxRate = 1.0f / rxInterval;
+
+    const float feedforwardSmoothFactor = pidGetFeedforwardSmoothFactor();
+    const float feedforwardJitterFactor = pidGetFeedforwardJitterFactor();
+    const float feedforwardTransitionFactor = pidGetFeedforwardTransitionFactor();
+    const float feedforwardBoostFactor = pidGetFeedforwardBoostFactor();
```

Comment:
Thanks for your ideas.
I've seen other bits of code with an enum list so that these things become feedforward.SmoothFactor etc.  But IDK about that in this case.
I use the full 'feedforward' name because a global search for 'feedforward' then finds every parameter related to feedforward.  I could use the abbreviation 'ff' but that finds 'off' and so on.

Reply:
For anything in the global scope I agree that it's a great idea to use *feedforward* as a prefix. For local variables in a function called `calculateFeedforward`, it feels a bit overkill, but not a big deal.

Reply:
And the advantage with creating a struct would be that the settings could be retrieved with a single function call, and perhaps later as an argument to the function, rather than needing a separate function and call for each value.

Reply:
Yes, IDK how to effectively set up a struct.  Perhaps we can resolve that later?